LIS3DH.device.lib.nut

```diff
@@ -0,0 +1,486 @@
+const LIS3DH_HPF_NORMAL_MODE            = 0x80;
+const LIS3DH_HPF_AUTORESET_ON_INTERRUPT = 0xC0;
+
+const LIS3DH_FIFO_BYPASS_MODE           = 0x00;
+const LIS3DH_FIFO_FIFO_MODE             = 0x40;
+const LIS3DH_FIFO_STREAM_MODE           = 0x80;
+const LIS3DH_FIFO_STREAM_TO_FIFO_MODE   = 0xC0;
+
+// Click Detection values
+const LIS3DH_SINGLE_CLICK  = 0x15;
+const LIS3DH_DOUBLE_CLICK  = 0x2A;
+
+const LIS3DH_MODE_NORMAL = 0x00;
+const LIS3DH_MODE_LOW_POWER = 0x01;
+const LIS3DH_MODE_HIGH_RESOLUTION = 0x02;
+
+const LIS3DH_ADC1 = 0x01;
+const LIS3DH_ADC2 = 0x02;
+const LIS3DH_ADC3 = 0x03;
+
+class LIS3DH {
+    static VERSION = "2.0.0";
+
+    // I2C information
+    _i2c = null;
+    _addr = null;
+    _mode = LIS3DH_MODE_NORMAL;
+
+    // The full-scale range (+/- _range G)
+    _range = null;
+
+    constructor(i2c, addr = 0x30) {
+        _i2c = i2c;
+        _addr = addr;
+
+        // Read the range + set _range property
+        getRange();
+    }
+    
+    // set default values for registers, read the current range and set _range
+    // (resets to state when first powered on)
+    function init() {
+        // Set default values for registers
+        _setReg(LIS3DH_CTRL_REG1, 0x07);
+        _setReg(LIS3DH_CTRL_REG2, 0x00);
+        _setReg(LIS3DH_CTRL_REG3, 0x00);
+        _setReg(LIS3DH_CTRL_REG4, 0x00);
+        _setReg(LIS3DH_CTRL_REG5, 0x00);
+        _setReg(LIS3DH_CTRL_REG6, 0x00);
+        _setReg(LIS3DH_INT1_CFG, 0x00);
+        _setReg(LIS3DH_INT1_THS, 0x00);
+        _setReg(LIS3DH_INT1_DURATION, 0x00);
+        _setReg(LIS3DH_CLICK_CFG, 0x00);
+        _setReg(LIS3DH_CLICK_SRC, 0x00);
+        _setReg(LIS3DH_CLICK_THS, 0x00);
+        _setReg(LIS3DH_TIME_LIMIT, 0x00);
+        _setReg(LIS3DH_TIME_LATENCY, 0x00);
+        _setReg(LIS3DH_TIME_WINDOW, 0x00);
+        _setReg(LIS3DH_FIFO_CTRL_REG, 0x00);
+
+        // Read the range + set _range property
+        getRange();
+    }
+
+    
+    // Enable the ADC. The ADC sampling frequency is the same as that of the ODR
+    // in LISxDH_CTRL_REG1. The input range is 0.8-1.6v and the
+    // data output is expressed in 2's complement left-aligned. The resolution is
+    // 8-bit in low-power mode 10 bits otherwise
+    function enableADC(state) {
+        _setRegBit(LIS3DH_TEMP_CFG_REG, 7, state);
+        _setRegBit(LIS3DH_CTRL_REG4, 7, state);
+    }
+    
+    // This method returns the current value of the ADC line. You must pass one of
+    // the lines (1, 2, or 3) or you will get an invalid return value. You must also
+    // have enabled the ADC by calling enableADC(true)
+    function readADC(ADC_line) {
+        local reg = (ADC_line * 2) + 6;
+        local read = _getMultiReg(reg, 2);
+        
+        local val = (((read[0] >> 6) | (read[1] << 2)) << 22) >> 22;
+        // Low power is only 8-bit resolution so we have to shift to get
+        // 8 bits and then scale appropriately
+        if(_mode == LIS3DH_MODE_LOW_POWER) {
+            val = (val>>2)/128.0;
```

Comment:
In fact this expression:

```
if(_mode == LIS3DH_MODE_LOW_POWER) {
    val = (val>>2)/128.0;
} else {
   val = val/512.0;
}
```

is equivalent to 

```
if(_mode == LIS3DH_MODE_LOW_POWER) {
   val = val/512.0;
} else {
   val = val/512.0;
}
```

which looks a little suspicious.